### ISSUE TITLE
VS Code (incorrectly) reports errors that CLI pyright does not

### ISSUE THREAD

**Issue by thisisrandy** [microsoft/pyright#2631]
**Describe the bug**
VS Code sometimes reports incorrect errors when CLI `pyright` is able to make a correct determination about (type) error-free code.

**To Reproduce**
Open the below code in VS Code and observe the reported errors. Note that there is some apparent non-determinism. E.g. in the second condition, edit to `elif foobar.bar == "ba" ...`. The error goes away but reappears if VS Code is restarted.

**Expected behavior**
VS Code error reporting should always agree with CLI `pyright`.

**Screenshots or Code**
The following code is a pared down version of the real code where this issue exhibits:

```python
from typing import Union
from dataclasses import dataclass, field


@dataclass
class Foo:
    foo: str


@dataclass
class Bar:
    bar: str
    baz: dict[str, str] = field(default_factory=dict)


FooBar = Union[Foo, Bar]


def do_stuff(foobar: FooBar) -> None:
    while foobar:
        if isinstance(foobar, Foo):
            pass
        elif foobar.bar == "bar" and "baz" in foobar.baz:
            return
```

Errors are reported at `"baz" in foobar.baz`. Hovering over `foobar` yields 

> Type of "baz" is unknown Pylance(reportUnknownMemberType)
> (parameter) foobar: Bar

and hovering over `baz` yields

> Type of "baz" is unknown Pylance(reportUnknownMemberType)
> Cannot access member "baz" for type "Foo"
>   Member "baz" is unknown Pylance(reportGeneralTypeIssues)
> (variable) baz: dict[str, str]

Screenshot as further proof:

![code error](https://user-images.githubusercontent.com/22106562/143794218-ddf6cce7-6641-43f4-87ad-7980dd658dc7.png)

Clearly pyright is able to determine that `foobar` is an instance of `Bar` in the condition for the section branch, because it doesn't complain about accessing `foobar.bar`, so it should likewise understand that `foobar.baz` is valid. Indeed, running `pyright` on the command line reports no errors.

**VS Code extension or command-line**
Pylance: v2021.11.2
Code: v1.62.3
OS: Ubuntu 20.04.3 LTS

**Additional context**
I'm not clear on whether this is, strictly speaking, a pyright issue. If not, please direct me to the appropriate forum.

**Comment by erictraut** [comment]
Thanks for reporting the problem. It has been nearly two weeks since the last release of pylance due to the Thanksgiving holiday in the U.S. Since the last pylance release, I've fixed the underlying bug that causes this behavior. The fix will be included in the next release of pylance (which is normally published each Wednesday).

**Comment by thisisrandy** [comment]
Fantastic. Thank you for the quick response. I'll close the issue as soon as I can confirm your fix.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.191, which I just published.

**Comment by thisisrandy** [comment]
Confirmed fixed in pylance v2021.12.0
